enhancements/proposals/1978-scylladbdatacenter-api/README.md

```diff
@@ -0,0 +1,764 @@
+
+Given following spec:
+```yaml
+spec:
+  topologyLabelSelector:
+    topology.kubernetes.io/region: us-east1
+  racks:
+  - name: a
+    topologyLabelSelector:
+      topology.kubernetes.io/zone: us-east1a
+```
+
+Results in rack `a` having `nodeAffinity`:
+```yaml
+nodeAffinity:
+  requiredDuringSchedulingIgnoredDuringExecution:
+    nodeSelectorTerms:
+      - matchExpressions:
+        - key: topology.kubernetes.io/region
+          operator: In
+          values:
+            - us-east1
+        - key: topology.kubernetes.io/zone
+          operator: In
+          values:
+          - us-east1a
+```
+
+### Uniform datacenter
+
+ScyllaDB recommendation is to have uniform racks. 
+I propose to extend the API so that the racks can be defined uniformly in a simplified manner, 
+while preserving the existing flexibility of having non-uniform racks by overriding it with dedicated specifications, if needed.
+
+#### Rack template
+
+New field `rackTemplate` on datacenter level, allows defining shared properties across all racks.
+Racks inherit these properties unless the same is specified on rack level, then overwritten value takes precedence.
+
+#### Node count
+
+A new field `nodes` will be added on the datacenter level. Scylla Operator will balance this number between all defined racks.
+Having a single place where the desired number of nodes is specified, enables us to have support for scale subresource,
+which brings users handy scaling operations using `kubectl scale` and paves a path towards available Kubernetes autoscalers.
+The downside of it is that these won't be able to add more racks, but provides a nice way of scaling for static topology clusters.
+
+Datacenter level `nodes` cannot be used together with `nodes` specified per rack. Webhook will prevent it.
+
+The balancing algorithm is iterative, adding/removing one node at a time in each iteration to ensure that racks stay as balanced as possible.
+
+```go
+func balanceNodesAcrossRacks(racks []*appsv1.StatefulSet, totalNodes int) []*appsv1.StatefulSet {
+    // Calculate the total number of existing nodes
+    existingNodes := 0
+    for _, rack := range racks {
+       existingNodes += rack.Replicas
+    }
+
+    if existingNodes == totalNodes {
+       // The desired number of nodes reached, do nothing.
+       return racks
+    }
+    
+    if existingNodes < totalNodes {
+       // Add a node to the least populated rack
+       minRackIndex := findSmallestStatefulSet(racks)
+       racks[minRackIndex].Replicas++
+       return balanceNodesAcrossRacks(racks, totalNodes)
+    }
+    
+    // Remove a node from biggest rack.
+    maxRackIndex := findBiggestStatefulSet(racks)
+    racks[maxRackIndex].Replicas--
+    return balanceNodesAcrossRacks(racks, totalNodes)
+}
+```
+
+New Condition will be added to the `status` of `v1alpha1.ScyllaDBDatacenter`, providing a simple way of noticing rack imbalance.
```

Comment:
what does imbalanced mean here? I assume even in a normal state when `numNodes%numRacks!=0` there will be up to `-1` difference, is that also to be considered to be imbalanced?

Reply:
my understanding is it just means state != desiredState (so in terms of the above func `sts != balanceNodesAcrossRacks(sts.racks, sts.nodes)` . But isn't that just progressing either way?

Reply:
after talking with Maciek on a call this was meant as node count or resources don't equal, even with n-1 difference - we may need to extend the nomenclature, but let's fist agreen on the distribution algorithm

Reply:
Removed